### DISCUSSION TITLE
how to get different instance of a WebAssembly module?

### DISCUSSION THREAD

**Discussion by Mr-shadow** [emscripten-core/emscripten#15873] [Q&A]
hi guys, 
   I watched this in MDN : "[A WebAssembly.Module object contains stateless WebAssembly code that has already been compiled by the browser — this can be efficiently shared with Workers, and instantiated multiple times.](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/Module)" , and  the book _[WebAssembly in Action](https://www.manning.com/books/webassembly-in-action)_  write by C.Gerard Gallant said that "Multiple instances of WebAssembly can share the same linear memory(ArrayBuffer)" in part 2.1.  And I want to know, will the memory of instance a be changed because of instance b while some module hold tow instance(instance a and instance b)?  I can't instantiate my module in this emcc version (3.0.1) now like this 
```
      const importObject = {    
        env: {
          __memory_base: 0,
        }
      };

      WebAssembly.instantiateStreaming(fetch("./side_module.wasm"), importObject).then(result => {
        const value = result.instance.exports._Increment(17);
        console.log(value.toString());
      });
```
whitch is come from the book _WebAssembly in Action_ base on emsdk version 1.38.45.

**Comment by sbc100** [comment]
In order for modules to share memory then need to import the memory from JS.   With emscripten, when you build with `-pthread` the JS code will generate a shared memory and then provide it as an import to all modules on all thread.

I would advise against attempted to run `WebAssembly.instantiateStreaming` yourself (especially if you want to make threads work), and instead rely on the emscripten-generate JavaScript code to take care of that for you.

**Reply by allan-bonadio** [reply]
Thanks for all your hard work getting C++ working on the browser!  I now have almost 6000 lines of C++ code, and it all works very well in the browser!  (here's a very early version: https://squish.tactileint.org )

This information about how to do wasm-workers needs to be written down somewhere.  WebAssembly.instantiateStreaming() is documented at MDN, 

https://developer.mozilla.org/en-US/docs/WebAssembly/JavaScript_interface/instantiateStreaming

so I read all about it and make my threads.  The Emscripten way of making wasm-workers isn't well documented at all.  So, I wander around, finding all the information I can, and raw info about WebAssembly.* has really helped me a lot.  

For my project, I just want to have several threads, or at least one, to do computations, sharing memory, and a few atomics, with the main thread.  Very simple.  Very little JS is needed running in the thread.  Just set it up, call in to C++, and run a loop forever.  This page tells me that I should make wasm-workers for my application, instead of pthreads:

https://emscripten.org/docs/api_reference/wasm_workers.html#pthreads-vs-wasm-workers-which-one-to-use

I've spent about a month trying to get wasm-workers to work; none of your emcc combinations seem to work.  How many other programmers have been lead down this dead end?

**Reply by sbc100** [reply]
Unless you have a good reason to, I would stick with using pthreads, which is the most long standing, well tested and standards based why to do threading in C/C++ with emscripten.  See https://emscripten.org/docs/porting/pthreads.html.

Unless you really know what you are doing I would avoid wasm workers for now.

**Reply by allan-bonadio** [reply]
What you just said: Why is it not said in the documentation?  I could have used that advice a month ago, instead of your misleading page.

**Reply by sbc100** [reply]
I guess you mean https://emscripten.org/docs/api_reference/wasm_workers.html#pthreads-vs-wasm-workers-which-one-to-use?   

Having read that perhaps my advice was a little strong.  I was under the impression that we were not recommending wasm workers very widely, but I may need to update my understanding.

If you want to use wasm workers it would work just fine.  We do test it.  I'm just personally more familiar with the cross platform pthreads API having worked with it for years, so to me it makes more sense to choose that option. 

Please file bugs against the wasm workers implementation if you find issues with it.
